File: src/cfg_ffi.erl
-module(cfg_ffi).

-export([
    new/0,
    add_vertex/2,
    add_edge/3,
    in_neighbors/2,
    out_neighbors/2
]).

new() ->
    digraph:new([cyclic]).

add_vertex(G, V) ->
    V = digraph:add_vertex(G, V),
    G.

add_edge(G, V1, V2) ->
    _ = digraph:add_edge(G, V1, V2),
    G.

in_neighbors(G, V) -> digraph:in_neighbours(G, V).

out_neighbors(G, V) -> digraph:out_neighbours(G, V).

File: src/cfg_ffi.mjs
import { List } from "./gleam.mjs";
import { Graph } from "@dagrejs/graphlib";

export function new_cfg() {
  return new Graph();
}

export function add_vertex(g, v) {
  g.setNode(v);
  return g;
}

export function add_edge(g, v1, v2) {
  g.setEdge(v1, v2);
  return g;
}

export function in_neighbors(g, v) {
  const neighbors = g.predecessors(v);
  return List.fromArray(neighbors);
}

export function out_neighbors(g, v) {
  const neighbors = g.successors(v);
  return List.fromArray(neighbors);
}
